packages/starlight/user-components/Tabs.astro

```diff
@@ -9,8 +9,70 @@ interface Props {
 const { syncKey } = Astro.props;
 const panelHtml = await Astro.slots.render('default');
 const { html, panels } = processPanels(panelHtml);
+
+/**
+ * Synced tabs are persisted across page using `localStorage`. The script used to restore the
+ * active tab for a given sync key has a few requirements:
+ *
+ * - The script should only be included when at least one set of synced tabs is present on the page.
+ * - The script should be inlined to avoid a flash of invalid active tab.
+ * - The script should only be included once per page.
+ *
+ * To do so, we keep track of whether the script has been rendered using a variable stored using
+ * `Astro.locals` which will be reset for each new page. The value is not typed on purpose to avoid
+ * Starlight users to get autocomplete for it.
+ *
+ * The restore script defines a custom element `starlight-tabs-restore` that will be included in
+ * each set of synced tabs to restore the active tab based on the persisted value using the
+ * `connectCallback` lifecycle method. To ensure this callback can access all tabs and panels for
+ * the current set of tabs, the script should be rendered before the tabs themselves.
+ */
+// @ts-expect-error - See above
+const { didRenderSyncedTabsRestoreScript } = Astro.locals;
+const isSynced = syncKey !== undefined;
+const shouldRenderSyncedTabsRestoreScript = isSynced && didRenderSyncedTabsRestoreScript !== true;
+
+if (isSynced) {
+	// @ts-expect-error - See above
+	Astro.locals.didRenderSyncedTabsRestoreScript = true;
+}
 ---
 
+{/* Inlined to avoid a flash of invalid active tab. */}
+{shouldRenderSyncedTabsRestoreScript && <script is:inline>
+	(() => {
+		class StarlightTabsRestore extends HTMLElement {
+			connectedCallback() {
+				const starlightTabs = this.closest('starlight-tabs');
+				if (!(starlightTabs instanceof HTMLElement) || typeof localStorage === 'undefined') return;
+				const syncKey = starlightTabs.dataset.syncKey;
+				if (!syncKey) return;
+				let label = '';
+				try {
+					const storedSyncedTabs = localStorage.getItem('starlight-synced-tabs') ?? '{}';
+					const persistedSyncedTabs = JSON.parse(storedSyncedTabs);
+					if (typeof persistedSyncedTabs !== 'object') return;
+					label = persistedSyncedTabs[syncKey];
+				} catch {}
+				if (!label) return;
+				const tabs = [...starlightTabs?.querySelectorAll('[role="tab"]')];
+				const tabIndexToRestore = tabs.findIndex((tab) => tab instanceof HTMLAnchorElement && tab.innerText === label);
```

Comment:
I feel like we did discuss this once in the past, but would `textContent` also work for syncing tabs? `innerText` is style aware so is [much slower](https://www.measurethat.net/Benchmarks/Show/3618/0/createtextnode-vs-textcontent-vs-innertext-vs-innerhtml) than `textContent` and in the context of this inline restore script _might_ cause style recalc in a similar way to what we just fixed in #2155.

Reply:
We did discuss it in the context of #1811 where we had to switch from `textContent` to `innerText`. I'll try it for sanity, but I think using `textContent` is this context would have the same issue of #1811.

Reply:
Ah, yes, thanks for remembering the context.

Maybe this test is too naive, but I just tried in the [Astro Embed docs](https://astro-embed.netlify.app/getting-started/) which uses your package manager tabs with icons. It looks like the difference is only in leading/trailing whitespace:

```
> tab.textContent
'   npm '
> tab.innerText
'npm'
```

In that case, I guess we could also solve it with `tab.textContent.trim()`?

Reply:
Forgot to mention, the recent changes are now all using trimmed `textContent` and I added some comments to make sure we remember why.